[Feature] Scale tests:FQM Latency & Queue Depth Performance, #384

ksaur opened the pull request:
## Summary

Perform scale testing for FQM/Latency Measurement

Closes https://github.com/NVIDIA/NVSentinel/issues/384

## Type of Change
- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 💥 Breaking change
- [x] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [x] Fault Management
- [ ] Health Monitors
- [ ] Janitor
- [ ] Other: ____________

## Testing
- [ ] Tests pass locally
- [ ] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [x] Documentation updated (if needed)
- [x] Ready for review


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * End-to-end scale test for measuring FQM latency, queue depth, and processing rates
  * Cluster management script to bulk reset cordoned nodes

* **Documentation**
  * Updated scale testing guide with new FQM latency objective and registry/setup note
  * Added detailed FQM latency & queue-depth results report and new Prometheus queries
  * Expanded event-generation and results documentation

* **Chores**
  * Made on-demand event-generation behavior configurable and updated ignore patterns for scale-test artifacts

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ksaur committed:
Scale tests: FQM latency testing support (#384)

- Add generateFatalGpuXidEventForFQM() with gpu-health-monitor agent for FQM latency tests
- Keep generateFatalGpuXidEvent() with event-generator agent for API/MongoDB tests (no accidental cordons)
- Add optional: true for EVENT_RATE to enable SIGUSR1 on-demand mode
- Add fqm-scale-test program for latency and queue depth measurement
- Add FQM latency test plan and reset script
- Add fqm-latency ConfigMap for SIGUSR1 mode

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
Add FQM Latency and Queue Depth test results for issue #384

Results from NVSentinel v0.4.0 scale testing on 1500-node cluster:

FQM Latency Measurement (150 nodes, 0-30s stagger):
- P90 latency: 28.15 seconds
- 100% success rate (150/150 nodes cordoned)

Queue Depth Measurement (150 nodes, 0-10s stagger):
- Peak queue depth: 138 nodes
- Time to clear: 149 seconds
- Processing rate: ~1 node/second
- 100% success rate

Includes:
- FQM_Latency_and_Queue_Depth_Results.md - detailed results document
- Raw CSV data files for latency and queue depth

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
Remove FQM_LATENCY_TEST_PLAN.md from repository

This file contains internal test notes and should not be in the public repo.
Added to .gitignore to prevent accidental commits.

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
Remove raw CSV data files from repository

Keep raw test data internal. Results are documented in
FQM_Latency_and_Queue_Depth_Results.md.

Added to .gitignore to prevent accidental commits.

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
feat(scale-tests): Lightweight FQM scale test tool + 750-node results

- Rewrote fqm-scale-test to use worker pool (50 workers default)
- Two-phase approach: blast all signals, then poll cordoned count
- Added progress output during Phase 1
- Updated reset script: scale FQM 0/1 instead of pod delete
- Added annotation cleanup to reset script
- 750-node (50%) test: 418.6s, 2.31 nodes/sec, 100% success

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
docs(scale-tests): Complete FQM latency results for 10%, 25%, 50% cluster

Lightweight test results (worker pool, no stagger):
- 10% (150 nodes): 64.5s, 3.33 nodes/sec
- 25% (375 nodes): 156.8s, 3.41 nodes/sec
- 50% (750 nodes): 418.6s, 2.31 nodes/sec

All tests 100% success rate.

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
docs(scale-tests): Complete FQM scale test results with graph

Phase 1 End-to-End Results (100% success):
- 10% (150 nodes): 64.5s, 3.33/sec
- 25% (375 nodes): 156.8s, 3.41/sec
- 50% (750 nodes): 307.1s, 3.49/sec

Prometheus Metrics:
- FQM Event Handling: P50 ~0.37s, P90 ~0.48s
- Platform Connector: ~5ms writes, zero queue backlog
- Bottleneck: FQM processing (not MongoDB)

Added event backlog overlay graph showing queue drain pattern

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
docs(scale-tests): Add FQM latency results and README summary

End-to-end results at 10%, 25%, 50% cluster failure:
- 100% cordoning success rate
- ~2-3 nodes/sec processing rate
- FQM bottleneck identified (~90% of total time)

Includes microbenchmarks, Prometheus metrics, and event backlog graph

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
docs(scale-tests): Update README and move event generation details

- Add FQM latency testing to objectives
- Add warning that event-generator image URL is placeholder only
- Move Event Distribution details to API_and_MongoDB_Results.md

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
updated graph

Signed-off-by: Karla Saur <ksaur@nvidia.com>

lalitadithya commented on tests/scale-tests/results/FQM_Latency_and_Queue_Depth_Results.md:
```diff
@@ -0,0 +1,163 @@
+- ✅ **~2.3 nodes/sec** at 50% load (750 nodes)
+- ✅ **Event handling latency** stays consistent: P50 ~0.37s, P90 ~0.48s
+- ✅ **Peak backlog scales linearly** with load: 104 → 256 → 534
+- ✅ **Bottleneck is FQM** — Platform Connector writes to MongoDB in ~5ms with zero queue backlog
```
I wouldn't say bottleneck. It has a queue and does one a time, so latency is expected to a certain extent. Can we rephrase this slightly so that the queue and one at a time is reflected?

ksaur replied:
fixed!

ksaur committed:
Address PR review comments for FQM scale tests

- Add license headers to main.go and reset-cordoned-nodes.sh
- Update go dependencies for security vulnerabilities (oauth2 v0.27.0, protobuf v1.33.0)
- Add godoc comment for QueueSnapshot struct
- Wrap errors with context using fmt.Errorf
- Remove deprecated rand.Seed() call (Go 1.20+ auto-seeds)
- Fix potential division by zero in avgRate calculation
- Refactor sendFatalEvent to use Kubernetes client-go API instead of kubectl exec
- Rephrase 'bottleneck' to 'sequential processing' in results doc

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur commented on tests/scale-tests/results/FQM_Latency_and_Queue_Depth_Results.md:
```diff
@@ -0,0 +1,163 @@
+# FQM Latency & Queue Depth Performance
+
+**Objective:** Measure FQM (Fault Quarantine Module) performance for cordoning nodes in response to fatal health events.
```
docs/METRICS.md:- [Fault Quarantine Module](#fault-quarantine)
fault-quarantine/Makefile:# Fault Quarantine Module Makefile

it is Module and not manager.

ksaur committed:
Address PR review comments and rerun scale tests

- Add baseline cordoned count to measure only new cordons per test run
- Guard against poll interval <= 0 (prevents panic)
- Refactor to use Kubernetes client-go instead of kubectl exec
- Suppress client-go klog noise, increase QPS/Burst for faster signal sending
- Add FQM and Platform Connector queries to PROMETHEUS_QUERIES.md
- Update test results with new data: 150/375/750 nodes at ~2.5 nodes/sec
- Regenerate backlog overlay graph with corrected timestamps
- Update key findings to reflect consistent ~2.5 nodes/sec rate

Signed-off-by: Karla Saur <ksaur@nvidia.com>

ksaur committed:
Merge main into scale-tests-issue-384

ksaur commented on tests/scale-tests/cmd/fqm-scale-test/main.go:
```diff
@@ -0,0 +1,420 @@
+	// Wait for polling to finish
+	<-pollDone
+
+	return snapshots
```
fixed

ksaur committed:
Update results to match Prometheus graph data

- Update peak queue values to Prometheus metrics: 142, 371, 688
- Fix green line time window to match 62.2s test duration
- Update key findings: ~2.5 nodes/sec consistent rate
- Update test date to December 5, 2025

Signed-off-by: Karla Saur <ksaur@nvidia.com>
